### DISCUSSION TITLE
Next 13 Include request and response in Page context, or some way of getting request and response for page

### DISCUSSION THREAD

**Discussion by SleeplessOne1917** [vercel/next.js#43362] [Ideas]
### Describe the feature you'd like to request

I'm currently using [next-session](https://github.com/hoangvvo/next-session#readme) to manage my sessions. 
In order to get the session, the request and response objects for a request are needed. In Next 12, this was done by using `getServerSideProps`, which had request and response objects in its context parameter. There is, to my knowledge, no replacement for getting the request and response objects in the new app folder routing system.

### Describe the solution you'd like

Currently I'd like to be able to write a server component like the following on "/app/create/page.tsx":
```typescript
const CreatePage = async ({ req, res }) => {
	const user = await getSessionUser(req, res);

	return <CreateIndexView username={user?.username} />;
};

export default CreatePage;
```

How exactly I get the request and response object isn't particularly important. These parameters could be added to pages, maybe layouts and templates, and any other type of special file deemed appropriate.

### Describe alternatives you've considered

Currently, according to the [file conventions docs page](https://beta.nextjs.org/docs/api-reference/file-conventions/page), the only 2 props are params and searchParams.

I've searched around the Next 13 docs to see if I could find a way to do it, but nothing sticks out to me. 

I've looked into using [iron-session](https://github.com/vvo/iron-session), but it's API didn't suit my needs and it also hasn't addressed the issue I'm presenting here.

**Comment by balazsorban44** [comment]
You have access to the request cookies and headers. See:

- https://beta.nextjs.org/docs/api-reference/cookies
- https://beta.nextjs.org/docs/api-reference/headers


`next-auth` (similar to `iron-session`) makes use of these primitives the following way:
https://github.com/nextauthjs/next-auth/blob/a307079e0fa4b4f1f1eae81140b8277773d4fa7e/packages/next-auth/src/next/index.ts#L143-L152

These are currently read-only, but we are working on an RFC for writing as well.

**Reply by arkmech** [reply]
headers and cookies aren't enough. Request object is needed. How would I handle subdomain logic?

**Reply by hubertlepicki** [reply]
@balazsorban44 I have very similar need, any update on the above? :)

**Reply by karlhorky** [reply]
@balazsorban44 is there a ballpark timeline on this RFC for writing to headers and cookies? Would be super useful primitives.

**Comment by hubertlepicki** [comment]
I have looked around Next.js code base and it appears that the Request is not exposed in any shape or form to the Server Components (in the /app). Instead, a `requestStore` is constructed [here](https://github.com/vercel/next.js/blob/canary/packages/next/server/app-render.tsx#L1909) that exposes only headers and cookies to the end programmer.

The simplest solution, which doesn't mean the best, would be to enrich the `requestStore` object with `req` property that would hold the `req` HTTP request object, and via that useful properties such as `body` or `url` would be available...

I could attempt to do it and provide a PR but again, this may be something authors want to explicitly avoid doing?

**Reply by karlhorky** [reply]
That link is now broken, I guess you mean [`packages/next/src/server/app-render.tsx`](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render.tsx)? I didn't find the `requestStore` that you mentioned though...

**Comment by tadaskay** [comment]
Thumbs up, I can't seem to find a way to handle a form POST in a server component (search params is not enough obviously).

**Comment by mrsawy** [comment]
I have the same problem  , the old approach was way more easier

**Comment by ndt080** [comment]
It is necessary to increase the priority of the problem.
It entails the inability to refresh jwt tokens on the server side.

**Comment by moofoo** [comment]
Was about to post this exact question. 

In a previous project I used [iron-session](https://github.com/vvo/iron-session) for authentication and storing a few user profile properties. I did authentication and re-routing in middleware, which translates fine to Next 13. The issue is getting the profile from the session and passing it to component props, which requires access to the Request and Response objects. I previously did this using getServerSideProps.
